Fully port FilterOpNodes to Rust

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit ports the FilterOpNodes pass to rust. This pass is exceedingly simple and just runs a filter function over all the op nodes and removes nodes that match the filter. However, the API for the class exposes that filter function interface as a user provided Python callable. So for the current pass we need to retain that python callback. This limits the absolute performance of this pass because we're bottlenecked by calling python.

Looking to the future, this commit adds a rust native method to DAGCircuit to perform this filtering with a rust predicate FnMut. It isn't leveraged by the python implementation because of layer mismatch for the efficient rust interface and Python working with `DAGOpNode` objects. A function using that interface is added to filter labeled nodes. In the preset pass manager we only use FilterOpNodes to remove nodes with a specific label (which is used to identify temporary barriers created by qiskit). In a follow up we should consider leveraging this new function to build a new pass specifically for this use case.

### Details and comments

Fixes #12263
Part of #12208

mtreinish committed:
Fully port FilterOpNodes to Rust

This commit ports the FilterOpNodes pass to rust. This pass is
exceedingly simple and just runs a filter function over all the op
nodes and removes nodes that match the filter. However, the API for
the class exposes that filter function interface as a user provided
Python callable. So for the current pass we need to retain that python
callback. This limits the absolute performance of this pass because
we're bottlenecked by calling python.

Looking to the future, this commit adds a rust native method to
DAGCircuit to perform this filtering with a rust predicate FnMut. It
isn't leveraged by the python implementation because of layer mismatch
for the efficient rust interface and Python working with `DAGOpNode`
objects. A function using that interface is added to filter labeled
nodes. In the preset pass manager we only use FilterOpNodes to remove
nodes with a specific label (which is used to identify temporary
barriers created by qiskit). In a follow up we should consider
leveraging this new function to build a new pass specifically for
this use case.

Fixes #12263
Part of #12208

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -5785,6 +5785,26 @@ impl DAGCircuit {
+    // Filter any nodes that don't match a given predicate function
+    pub fn filter_op_nodes<F, E>(&mut self, mut predicate: F) -> Result<(), E>
+    where
+        F: FnMut(&PackedInstruction) -> Result<bool, E>,
```
I wonder if we should allow the callable to be fallible at all. I know we have often supported this in Rustworkx core's API to make it easier to use with the wild world of Python. But in this case, maybe it's okay to require the user to give us a function that isn't expected to fail.

Otherwise, we'll be asking them to define `Infallible` result types probably more often than not.

mtreinish replied:
I'm fine with dropping the result here, I added it because when I originally wrote this I intended to use this method for the python function. But because that wasn't possible because of different interfaces between python and rust. So I agree this probably doesn't need to be fallible if we're not calling this with a python callback. I'll remove the result as it'll make it easier to work with.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13052/commits/5e00d3c416172a82d63e0ca378ef2465292edd7e

kevinhartman commented on crates/accelerate/src/filter_op_nodes.rs:
```diff
@@ -0,0 +1,65 @@
+///     dag (DAGCircuit): The dag circuit to filter the ops from
+///     label (str): The label to filter nodes on
+#[pyfunction]
+pub fn filter_labelled_op(dag: &mut DAGCircuit, label: String) {
```
It sounds to me like `labeled` is the preferred American spelling but `labelled` is preferentially in British English.

mtreinish replied:
Hah @jakelishman will be proud of me because this looks normal to me. :gb:

I can update the spelling to use the conventional American spelling.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13052/commits/1a99b80736f072f9e90d07a80c1877a712ee3a1d

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -5785,6 +5785,26 @@ impl DAGCircuit {
     }
 
+    // Filter any nodes that don't match a given predicate function
+    pub fn filter_op_nodes<F, E>(&mut self, mut predicate: F) -> Result<(), E>
```
I think the term "filter" makes more sense in the context of a transpiler pass / pipeline than it does directly on `DAGCircuit`, but I'm not sure I can think of a better name. Maybe something like `remove_op_nodes_by_filter`.

I'm fine with this comment being closed if anyone (yourself included) prefers the current name.

mtreinish replied:
Personally I thought this was pretty clear, using filter as a verb typically means to remove like this. But I'll let others chime in here, because my thoughts/reading on things aren't always clear to others.

kevinhartman replied:
I think your thoughts on it are valid. The name just makes me think of an iterator kind of situation i.e. if you gave me just the method name I would not guess that it took a `&mut self`. But, it does. So it's up to personal preference I suppose.

jakelishman replied:
fwiw, compare Rust's `Iterator::filter` (new iterator containing only matching items) and Python's builtin `filter` (same).  Changing to `remove_op_nodes_by_filter` would invert the polarity of the `predicate` compared to the existing code and would probably conflict with Rust and Python builtins.

jakelishman replied:
and for the `&self` vs `&mut self` point, both Rust and Python's methods are mutations on the incoming iterator, which is consistent with `&mut self` (well, strictly `self`, but the analogy starts breaking down there

kevinhartman replied:
Heh, the analogy breaking down there is the problem I have with this, since we're not consuming the DAG, so it's no longer a filter of some pipelined data flow, but instead removals from a sitting pool.

But, I'm fine with keeping the name as is since I'm clearly in the minority on this 🙂.

kevinhartman reviewed (commented):
Looks good, just a few minor comments.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into rust-filter-op-nodes

mtreinish committed:
Make filter_op_nodes() infallible

The filter_op_nodes() method originally returned a Result<()> to handle
a predicate that was fallible. This was because the original intent for
the method was to use it with Python callbacks in the predicate. But
because of differences between the rust API and the Python API this
wasn't feasible as was originally planned. So this Result<()> return
wasn't used anymore. This commit reworks it to make the
filter_op_nodes() infallible and the predicate a user provides also only
returns `bool` and not `Result<bool>`.

mtreinish committed:
Rename filter_labelled_op to filter_labeled_op

mtreinish committed:
Merge remote-tracking branch 'origin/main' into rust-filter-op-nodes

kevinhartman reviewed (approved):
Looks good. Thanks for the changes!
